tools/ctl/ctl.c

```diff
@@ -540,6 +567,30 @@ int main(int argc, char *argv[])
 		}
 	}
 
+	/* open output file */
+	if (output_file) {
+		ctl_data->out_fd = open(output_file, O_CREAT | O_TRUNC | O_RDWR,
+					0600);
+		if (ctl_data->out_fd <= 0) {
+			fprintf(stderr, "error: %s\n", strerror(errno));
+			goto in_fd_close;
+		}
+	}
+
+	/* Just print the ABI header if requested */
+	if (ctl_data->print_abi_header) {
+		ctl_data->ctrl_size = sizeof(struct sof_abi_hdr);
+		buffer_alloc(ctl_data);
+		header_init(ctl_data);
+		hdr = (struct sof_abi_hdr *)
+			&ctl_data->buffer[BUFFER_ABI_OFFSET];
+		hdr->size = ctl_data->print_abi_size;
+		ctl_data->buffer[BUFFER_SIZE_OFFSET] = ctl_data->ctrl_size;
+		ctl_dump(ctl_data);
+		buffer_free(ctl_data);
+		goto struct_free;
```

Comment:
goto out_fd_close;

Reply:
The sequence here is changed, so we need revisit them for each item below, e.g.
```
	/* The control need to be defined. */
	if (!ctl_data->cname) {
		fprintf(stderr, "Error: No control was requested.\n");
		usage(argv[0]);
--		goto struct_free;
++		goto out_fd_close;
	}

	/* open input file */
	if (input_file) {
		ctl_data->in_fd = open(input_file, O_RDONLY);
		if (ctl_data->in_fd <= 0) {
			fprintf(stderr, "error: %s\n", strerror(errno));
--			goto struct_free;
++		        goto out_fd_close;
		}
	}

	/* set up ctl_elem, allocate buffers */
	ret = ctl_setup(ctl_data);
	if (ret < 0) {
		fprintf(stderr, "Error: ctl_data setup failed, ret:%d", ret);
--		goto out_fd_close;
++		goto in_fd_close;
	}
```

and change at the end of the function like this:

```
data_free:
	ret = ctl_free(ctl_data);
in_fd_close:
	if (ctl_data->in_fd)
		close(ctl_data->in_fd);
out_fd_close:
	if (ctl_data->out_fd)
		close(ctl_data->out_fd);
struct_free:
	free(ctl_data);
```